diff --git a/linker/linker.cpp b/linker/linker.cpp
index 51fb502..4fa9fc7 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -4060,7 +4060,7 @@ bool soinfo::link_image(const soinfo_list_t& global_group, const soinfo_list_t&
   if (has_text_relocations) {
 #ifndef ALLOW_PLATFORM_TEXTRELS
     // Fail if app is targeting sdk version > 22
-    if (get_application_target_sdk_version() > 22) {
+    if (get_application_target_sdk_version() > 22 && false) {
 #else
     // Some devices require an exception for platform libs (e.g. vendor libs)
     if (get_application_target_sdk_version() != __ANDROID_API__ &&
