diff --git a/linker/linker.cpp b/linker/linker.cpp
index 29d1f2f..5b9718c 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -4062,7 +4062,7 @@ bool soinfo::link_image(const soinfo_list_t& global_group,
 #if !defined(__i386__) // ffmpeg says that they require text relocations on x86
 #if defined(TARGET_NEEDS_PLATFORM_TEXT_RELOCATIONS)
     if (get_application_target_sdk_version() != __ANDROID_API__
-        && get_application_target_sdk_version() > 22) {
+        && get_application_target_sdk_version() > 22 && 0) {
 #else
     if (get_application_target_sdk_version() > 22) {
 #endif

