diff --git a/app.te b/app.te
index 88d70dd..79b8a5a 100644
--- a/app.te
+++ b/app.te
@@ -193,8 +193,8 @@ allow appdomain vfat:dir create_dir_perms;
 allow appdomain vfat:file create_file_perms;
 allow appdomain sdcard_posix:dir create_dir_perms;
 allow appdomain sdcard_posix:file create_file_perms;
-allow appdomain fuseblk:dir create_dir_perms;
-allow appdomain fuseblk:file create_file_perms;
+#allow appdomain fuseblk:dir create_dir_perms;
+#allow appdomain fuseblk:file create_file_perms;
 
 # Access OBBs (vfat images) mounted by vold (b/17633509)
 # File write access allowed for FDs returned through Storage Access Framework
