diff --git a/updater/updater.cpp b/updater/updater.cpp
index e956dd5..e0c48d4 100644
--- a/updater/updater.cpp
+++ b/updater/updater.cpp
@@ -36,6 +36,9 @@
 // (Note it's "updateR-script", not the older "update-script".)
 #define SCRIPT_NAME "META-INF/com/google/android/updater-script"
 
+// golden/janice unification
+#define UNIFIED_BUILD
+
 extern bool have_eio_error;
 
 struct selabel_handle *sehandle;
@@ -87,11 +90,33 @@ int main(int argc, char** argv) {
     }
     ota_io_init(&za);
 
+#ifndef UNIFIED_BUILD
     const ZipEntry* script_entry = mzFindZipEntry(&za, SCRIPT_NAME);
     if (script_entry == NULL) {
         printf("failed to find %s in %s\n", SCRIPT_NAME, package_filename);
         return 4;
     }
+#else
+	// deviceName - the name of the device (eg. golden)
+	char deviceName[PROPERTY_VALUE_MAX];
+	property_get("ro.product.device", deviceName, "");
+	
+	// scriptName - due to golden and janice having different partition layouts
+	// (eg. /system is mmcblk0p22 on golden and mmcblk0p3 on janice), we need to have
+	// different updater scripts for the two devices.
+    char scriptName[PROPERTY_VALUE_MAX];
+	if (deviceName == "janice") scriptName = "META-INF/com/google/android/updater-script-janice"; // janice will use updater-script-janice
+	else scriptName = "META-INF/com/google/android/updater-script"; // while golden will use updater-script
+	
+	// print out the device name for debug purposes
+	printf("%s device detected" % deviceName);
+	
+	const ZipEntry* script_entry = mzFindZipEntry(&za, scriptName);
+    if (script_entry == NULL) {
+        printf("failed to find %s in %s\n", scriptName, package_filename);
+        return 4;
+    }
+#endif
 
     char* script = reinterpret_cast<char*>(malloc(script_entry->uncompLen+1));
     if (!mzReadZipEntry(&za, script_entry, script, script_entry->uncompLen)) {
