diff --git a/linker/linker.cpp b/linker/linker.cpp
index d1f3469..04802c9 100644
--- a/linker/linker.cpp
+++ b/linker/linker.cpp
@@ -4084,9 +4084,6 @@ bool soinfo::LinkImage(const android_dlextinfo* extinfo) {
 #endif
     // Make segments writable to allow text relocations to work properly. We will later call
     // phdr_table_protect_segments() after all of them are applied.
-    DL_WARN("%s has text relocations. This is wasting memory and prevents "
-            "security hardening. Please fix.", get_realpath());
-    add_dlwarning(get_realpath(), "text relocations");
     if (phdr_table_unprotect_segments(phdr, phnum, load_bias) < 0) {
       DL_ERR("can't unprotect loadable segments for \"%s\": %s",
              get_realpath(), strerror(errno));
