Subject: [PATCH] busybox: Fix build

this is a workaround for clang errors

diff --git a/Android.mk b/Android.mk
index d0c6b9a..b715f56 100644
--- a/Android.mk
+++ b/Android.mk
@@ -43,7 +43,7 @@
 	@rm -f $(shell find $(abspath $(call intermediates-dir-for,EXECUTABLES,busybox)) -name "*.o")
 	@mkdir -p $(@D)
 	@cat $^ > $@ && echo "CONFIG_CROSS_COMPILER_PREFIX=\"$(BUSYBOX_CROSS_COMPILER_PREFIX)\"" >> $@
-	+make -C $(BB_PATH) prepare O=$(@D)
+	make -C $(BB_PATH) prepare O=$(@D)
 
 busybox_prepare_minimal := $(bb_gen)/minimal/.config
 $(busybox_prepare_minimal): $(BB_PATH)/busybox-minimal.config
@@ -52,7 +52,7 @@
 	@rm -f $(shell find $(abspath $(call intermediates-dir-for,STATIC_LIBRARIES,libbusybox)) -name "*.o")
 	@mkdir -p $(@D)
 	@cat $^ > $@ && echo "CONFIG_CROSS_COMPILER_PREFIX=\"$(BUSYBOX_CROSS_COMPILER_PREFIX)\"" >> $@
-	+make -C $(BB_PATH) prepare O=$(@D)
+	make -C $(BB_PATH) prepare O=$(@D)
 
 
 #####################################################################
diff --git a/Android.mk b/Android.mk
index d0c6b9a..b715f57 100644
--- a/Android.mk
+++ b/Android.mk
@@ -93,7 +93,7 @@ BUSYBOX_C_INCLUDES = \
 	$(BB_PATH)/android/librpc
 
 BUSYBOX_CFLAGS = \
-	-Werror=implicit -Wno-clobbered \
+	-Werror=implicit -Wno-clobbered -Wno-format-security \
 	-DNDEBUG \
 	-DANDROID \
 	-fno-strict-aliasing \
