diff --git a/services/core/java/com/android/server/BatteryService.java b/services/core/java/com/android/server/BatteryService.java
index 3702105..ff6f6bb 100644
--- a/services/core/java/com/android/server/BatteryService.java
+++ b/services/core/java/com/android/server/BatteryService.java
@@ -966,7 +966,7 @@
         }
 
         private boolean isHvdcpPresent() {
-            File mChargerTypeFile = new File("/sys/class/power_supply/usb/type");
+            File mChargerTypeFile = new File("/sys/class/power_supply/ab8500_usb/type");
             FileReader fileReader;
             BufferedReader br;
             String type;
