diff --git a/InCallUI/src/com/android/incallui/CallList.java b/InCallUI/src/com/android/incallui/CallList.java
index 5f0f36f..bb89d10 100644
--- a/InCallUI/src/com/android/incallui/CallList.java
+++ b/InCallUI/src/com/android/incallui/CallList.java
@@ -52,8 +52,8 @@public class CallList implements InCallPhoneListener {
 public class CallList {
 
     private static final int DISCONNECTED_CALL_SHORT_TIMEOUT_MS = 200;
-    private static final int DISCONNECTED_CALL_MEDIUM_TIMEOUT_MS = 2000;
-    private static final int DISCONNECTED_CALL_LONG_TIMEOUT_MS = 5000;
+    private static final int DISCONNECTED_CALL_MEDIUM_TIMEOUT_MS = 500;
+    private static final int DISCONNECTED_CALL_LONG_TIMEOUT_MS = 1000;
 
     private static final int EVENT_DISCONNECTED_TIMEOUT = 1;
     private static final int EVENT_NOTIFY_CHANGE = 2;
