Subject: [PATCH] sad selinux 2002

diff --git a/init/service.cpp b/init/service.cpp
index ef59d7e..af8791d 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -371,12 +371,6 @@
             scon = ret_scon;
             free(ret_scon);
         }
-        if (rc == 0 && scon == mycon) {
-            ERROR("Service %s does not have a SELinux domain defined.\n", name_.c_str());
-            free(mycon);
-            free(fcon);
-            return false;
-        }
         free(mycon);
         free(fcon);
         if (rc < 0) {
