diff --git a/cryptfs.c b/cryptfs.c
old mode 100644
new mode 100755
index c697af6..85342a4
--- a/cryptfs.c
+++ b/cryptfs.c
@@ -111,6 +111,8 @@ static struct crypt_persist_data *persist_data = NULL;
 
 static int previous_type;
 
+static bool legacy_dm_crypt = false;
+
 #ifdef CONFIG_HW_DISK_ENCRYPTION
 static int scrypt_keymaster(const char *passwd, const unsigned char *salt,
                             unsigned char *ikey, void *params);
