Fixes offline charging
diff --git a/init/init.cpp b/init/init.cpp
index 9992b47..84e644d 100755
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -87,6 +87,10 @@ static time_t process_needs_restart;
 
 const char *ENV[32];
 
+#ifdef STE_HARDWARE
+static unsigned lpm_bootmode = 0;
+#endif
+
 bool waiting_for_exec = false;
 
 static int epoll_fd = -1;
@@ -334,6 +338,12 @@ static void import_kernel_nv(const std::string& key, const std::string& value, b
 
     if (key == "qemu") {
         strlcpy(qemu, value.c_str(), sizeof(qemu));
+#ifdef STE_HARDWARE
+     } else if (key == "lpm_boot") {
+        if (value == "1") {
+            lpm_bootmode = 1;
+        }
+#endif
     } else if (key == BOARD_CHARGING_CMDLINE_NAME) {
         strlcpy(battchg_pause, value.c_str(), sizeof(battchg_pause));
     } else if (android::base::StartsWith(key, "androidboot.")) {
@@ -532,6 +560,9 @@ static void selinux_initialize(bool in_kernel_domain) {
 }
 
 static int charging_mode_booting(void) {
+#ifdef STE_HARDWARE
+    return lpm_bootmode;
+#else
 #ifndef BOARD_CHARGING_MODE_BOOTING_LPM
     return 0;
 #else
@@ -549,6 +580,7 @@ static int charging_mode_booting(void) {
     close(f);
     return ('1' == cmb);
 #endif
+#endif
 }
 
 int main(int argc, char** argv) {
