diff --git a/services/core/jni/com_android_server_location_GnssLocationProvider.cpp b/services/core/jni/com_android_server_location_GnssLocationProvider.cpp
index c6cd714..8ab3a5c 100644
--- a/services/core/jni/com_android_server_location_GnssLocationProvider.cpp
+++ b/services/core/jni/com_android_server_location_GnssLocationProvider.cpp
@@ -876,15 +876,15 @@ static void android_location_GnssLocationProvider_agps_data_conn_open(
     }
 
     const char *apnStr = env->GetStringUTFChars(apn, NULL);
-
+/*
     size_t interface_size = sAGpsInterface->size;
     if (interface_size == sizeof(AGpsInterface)) {
         sAGpsInterface->data_conn_open_with_apn_ip_type(apnStr, apnIpType);
-    } else if (interface_size == sizeof(AGpsInterface_v1)) {
+    } else if (interface_size == sizeof(AGpsInterface_v1)) {*/
         sAGpsInterface->data_conn_open(apnStr);
-    } else {
+    /*} else {
         ALOGE("Invalid size of AGpsInterface found: %zd.", interface_size);
-    }
+    }*/
 
     env->ReleaseStringUTFChars(apn, apnStr);
 }
