diff --git a/minui/graphics.cpp b/minui/graphics.cpp
index 1f25531..affce59 100644
--- a/minui/graphics.cpp
+++ b/minui/graphics.cpp
@@ -425,7 +425,7 @@ void gr_flip() {
     gr_draw = gr_backend->flip(gr_backend);
 }
 
-int gr_init(void)
+int gr_init_real(void)
 {
     gr_init_font();
 
@@ -459,6 +459,13 @@ int gr_init(void)
     return 0;
 }
 
+int gr_init(void)
+{
+    gr_init_real();
+    gr_init_real();
+    return 0;
+}
+
 void gr_exit(void)
 {
     gr_backend->exit(gr_backend);
