diff --git a/include/binder/Parcel.h b/include/binder/Parcel.h
index eb714db..679f6ed 100644
--- a/include/binder/Parcel.h
+++ b/include/binder/Parcel.h
@@ -477,7 +495,8 @@ private:
             return val.flatten(buffer, size, fds, count);
         }
         virtual status_t unflatten(void const* buffer, size_t size, int const* fds, size_t count) {
-            status_t err = const_cast<Flattenable<T>&>(val).unflatten(buffer, size, fds, count);
+            //status_t err = const_cast<Flattenable<T>&>(val).unflatten(buffer, size, fds, count);
+            status_t err = const_cast<Flattenable&>(val).unflatten(buffer, size, fds, count); // STE
             if (err != NO_ERROR) {
                 for (size_t i = 0; i < count; i++) {
                     close(fds[i]);
